### ISSUE TITLE
ec2:TerminateInstances already included in execution role policy

### ISSUE THREAD

**Issue by tadeasbronis** [aws-cloudformation/aws-cloudformation-resource-providers-awsutilities-commandrunner#24]
I tried to use in our company, but with my restricted access I ended up with an error: `You do not have permissions to make the TerminateInstances API call. Please try again with the necessary permissions.`

I checked your code and found out, that the error is thrown because You are trying to add `ec2:TerminateInstances` in CreateHandler.java again. But `ec2:TerminateInstances` permissions are already added in execution role policy. Can You please delete that of the code? Thanks.

**Comment by maslick** [comment]
Same issue here...

**Comment by maslick** [comment]
He's not adding the policy, but rather simulates if the requester principal (IAM user or role) has enough rights to ``ec2:TerminateInstances``. I have no idea why this check is needed...

**Comment by maslick** [comment]
This PR #30 solves the issue

**Comment by jlongo62** [comment]
So if this thing does not work how do we get maslick's changes integrated into a release ?

**Comment by maslick** [comment]
@jlongo62 i don't think this repo is maintained...

**Comment by jlongo62** [comment]
are you using this ? do you have it working ?

**Comment by maslick** [comment]
@jlongo62 yes, we are using this in production

**Comment by jlongo62** [comment]
what is the workaround ?

**Comment by maslick** [comment]
@jlongo62 see #30

**Comment by maslick** [comment]
@jlongo62 so what? use the feature branch and follow the instructions

**Comment by jlongo62** [comment]
I can do the git and get the pull request.
But do you have a script to install the dependencies on AWS cloudshell to be able to process ./scripts/build.sh

**Comment by maslick** [comment]
@jlongo62 you need:

- Java 1.8+
- Maven 3+
- cfn cli
- AWS CLI v2

```
# Install cfn cli
pip install cloudformation-cli cloudformation-cli-java-plugin cloudformation-cli-go-plugin cloudformation-cli-python-plugin cloudformation-cli-typescript-plugin

# Clone maslick's fork of CommandRunner, checkout fix-terminate-instances branch
git clone git@github.com:maslick/aws-cloudformation-resource-providers-awsutilities-commandrunner.git

э

**Comment by jlongo62** [comment]
Any chance of you publishing a release for this in your repository ?

**Comment by jlongo62** [comment]
Ok. I have spent an entire day trying to build a machine to build this java code (that does something that should already be in the CF box AND that I would not need if CF components were not built by people who hate us) only to get this.
Any chance of you just publishing a release in your repository?

```
[INFO] --- maven-resources-plugin:2.4:resources (default-resources) @ awsutility-cloudformation-commandrunner-handler ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ awsutility-cloudformation-commandrunner-handler ---
[INFO] Changes detected - recompiling the module! see the AWS CL
[INFO] Compiling 11 source files to /home/ec2-user/aws-cloudformation-resource-providers-awsutilities-commandrunner/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] No compiler is provided in this environment. Perhaps you are running on a JRE rather than a JDK?
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.110 s
[INFO] Finished at: 2021-12-01T20:59:44Z
[

```

**Comment by jlongo62** [comment]
Now I cannot get the aws commands to work. I have a simple resource straight from the docs:
```
    GetransitGatewayRouteTableId:
      Type: AWSUtility::CloudFormation::CommandRunner
      DependsOn: [VPNConnection]
      Properties: 
        Role: InfrastructureManagement
        LogGroup: "awsutility-cloudformation-commandrunner-logs"
        Command: 'aws s3 ls | sed -n 1p | cut -d " " -f3 > /command-output.txt'    
```
It returns the error:

> Resource handler returned message: "1 validation error detected: Value at 'value' failed to satisfy constraint: Member must have length greater than or equal to 1. (Service: AWSSimpleSystemsManagement; Status Code: 400; Error Code: ValidationException; Request ID: 915da8da-06f2-49f0-80da-0b6d7140a73b; Proxy: null)" (RequestToken: 9410d74b-a60e-f604-d7d5-6344199b7de3, HandlerErrorCode: GeneralServiceException)

**Comment by maslick** [comment]
@jlongo62 you should also consider this #32 

Can you use this template?
```yml
AWSTemplateFormatVersion: 2010-09-09
Description: AWS CLI version


Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: !Sub "/command-runner/${AWS::StackName}/"
      RetentionInDays: 14

  Role:
    Type: AWS::IAM::Role
    Properties:
      Description: "Role assumed by Command Runnner"
      MaxSessionDuration: 14400
      Path: "/"
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service: ec2.amazonaws.com
            Action: sts:AssumeRole
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy

  InstanceProfile:
    Type: AWS::IAM::InstanceProfile
    Properties:
      Path: /
      Roles:
        - !Ref Role

  CommandRunner:
    Type: AWSUtility::CloudFormation::CommandRunner
    Properties:
      LogGroup: !Ref LogGroup
      Role: !Ref InstanceProfile
      SubnetId: YOUR_SUBNET_ID
      Command: |
        aws --version 2>&1 \
          | tail -1 \
          | head -n 1 \
          > /command-output.txt


Outputs:
  AwsCliVersion:
    Description: AWS CLI version
    Value: !GetAtt CommandRunner.Output
```

Check the logs from ``/command-runner/${AWS::StackName}/`` log group

**Comment by maslick** [comment]
@jlongo62 you are getting this error, probably because your role doesn't have an instance profile. An Instance profile is created automatically when you create an EC2 instance from the console. ``CommandRunner`` is expecting the Instance profile, and not a plain Role in the ``Role`` property. This field is counterintuitive and should be rather renamed to ``InstanceProfile``.
You are absolutely right that if you do not specify any subnet your account must have a default VPC and the region should support ``t3.micro`` instance type.

**Comment by maslick** [comment]
BTW we are using [this branch](https://github.com/maslick/aws-cloudformation-resource-providers-awsutilities-commandrunner/tree/fix-null-sg) (from #32) in production, since we don't have a default VPC, and we need to run CommandRunner in a private subnet.

**Comment by jlongo62** [comment]
_Credit to maslick; his code works...._

Attached is:

1. A shell script will build and deploy commandRunner using CloudShell.
2. The YAML to make an assumable role to run CloudShell undert is also included (I pulled it from another project, handy if you do not have admin rights)
3. A non-trivial, but simple test CF to validate commandRunner can use the AWS CLI.

[commandrunner.zip](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-awsutilities-commandrunner/files/7656797/commandrunner.zip)

maslick: feel freeto incorporate this in your repository.
